controller/api/destination/external-workload/endpoints_controller.go

```diff
@@ -0,0 +1,513 @@
+					ec.log.Infof("%s acquired lease", hostname)
+				}
+			},
+		},
+	}
+
+	return ec, nil
+}
+
+// Start will run the endpoint manager's processing loop and leader elector.
+//
+// The function will spawn two background tasks; one to run the leader elector
+// client that and one that will process updates applied by the informer
+// callbacks.
+func (ec *EndpointsController) Start() {
+	// Create a parent context that will be used by leader elector to gracefully
+	// shutdown.
+	//
+	// When cancelled (either through cancel function or by having its done
+	// channel closed), the leader elector will release the lease and stop its
+	// execution.
+	ctx, cancel := context.WithCancel(context.Background())
+
+	go func() {
+		for {
+			// Block until a lease is acquired or a lease has been released
+			leaderelection.RunOrDie(ctx, ec.lec)
+			// If the context has been cancelled, exit the function, otherwise
+			// continue spinning.
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+		}
+	}()
+
+	// Start a background task to process updates. When a shutdown signal is
+	// received over the manager's stop channel, it is propagated to the elector
+	// through the context object to ensure the elector task does not leak.
+	go func() {
+		for {
+			select {
+			case update := <-ec.updates.queue:
+				elapsed := time.Since(update.enqueTime).Seconds()
+				ec.updates.queueLatency.Observe(elapsed)
+				ec.updates.queueLength.Dec()
+
+				ec.processUpdate(update.item)
+			case <-ec.stop:
+				ec.log.Info("Received shutdown signal")
+				// Propagate shutdown to elector through the context.
+				cancel()
+				return
+			}
+		}
+	}()
+}
+
+// enqueueUpdate will enqueue a service key to the updates buffer to be
+// processed by the endpoint manager. When a write to the buffer blocks, the
+// update is dropped.
+func (ec *EndpointsController) enqueueUpdate(key string) {
+	update := queueUpdate{
+		item:      key,
+		enqueTime: time.Now(),
+	}
+	select {
+	case ec.updates.queue <- update:
+		// Update successfully enqueued.
+		ec.updates.queueLength.Inc()
+		ec.updates.queueUpdates.Inc()
+	default:
+		// Drop update
+		ec.updates.queueDrops.Inc()
+		ec.log.Debugf("External endpoint manager queue is full; dropping update for %s", update.item)
+		return
+	}
+}
+
+func (ec *EndpointsController) processUpdate(update string) {
+	// TODO (matei): reconciliation logic
+	// TODO (matei): track how long an item takes to be processed
+	ec.log.Infof("Received %s", update)
+}
+
+// getSvcMembership accepts a pointer to an external workload resource and
+// returns a set of service keys (<namespace>/<name>). The set includes all
+// services local to the workload's namespace that match the workload.
+func (ec *EndpointsController) getSvcMembership(workload *ewv1alpha1.ExternalWorkload) (map[string]struct{}, error) {
+	set := make(map[string]struct{})
+	services, err := ec.k8sAPI.Svc().Lister().Services(workload.Namespace).List(labels.Everything())
```

Comment:
As in pass the labels from the workload? That might not fully work since services and workloads can have a different set of labels. For example:

```
apiVersion: v1
kind: Service
metadata:
  name: my-svc
spec:
  selector:
    app=my-app
---
apiVersion: workload.linkerd.io/v1alpha1
kind: ExternalWorkload
metadata:
  name: my-workload
  labels:
    app=my-app
```

This would be a valid example. Since we end up checking the service's _selector_ we can't restrict our list to only get services that match the workload's labels.